### PR TITLE
Deep copy MLP

### PR DESCRIPTION
**Pull request by wulele2**
I admire your open source code very much, so I learned a lot. In Pytorch, **nn.ModuleList[module for I in range(N)]** is essentially a shallow copy of module. In order to use different modules at different layers, copy is required to complete deep copy. Therefore, I modified it with copy. Maybe I misunderstood it wrong. Welcome your comments.

### CODE REVIEW

**Comment by wulele2** [comment]
`'''
nn.ModuleList是浅拷贝，即共享权重；
要想支持深拷贝需要copy模块
'''
import os
import copy
import torch
import torch.nn as nn
import torch.optim as optim

os.environ['CUDA_VISIBLE_DEVICES'] = '1'

\#创建一个线性层
layer = nn.Linear(256,4)
\#深拷贝
out1 = nn.ModuleList([copy.deepcopy(layer) for _ in range(6)])
\# 查看深拷贝方式的layer层初始化参数
print('深拷贝的不同层的初始化权重：\n',out1[0].weight.data, '\n', out1[1].weight.data)

\#验证深拷贝
print('验证深拷贝：')
input = torch.randn(1,256)
label = torch.ones((1,4), dtype= torch.float32)
\# 取出第一层layer来作为net
net1 = out1[0]
optimizer1 = optim.SGD(net1.parameters(), lr=0.1)
criterion1 = nn.SmoothL1Loss()

\# 迭代
for i in range(1):
    optimizer1.zero_grad()
    res1 = net1(input)
    loss1 = criterion1(res1, label)
    loss1.backward()
    optimizer1.step()
    print('更新后的网络参数：\n',net1.weight.data)
    print('其余层网络参数：\n',out1[1].weight.data)

print('#-------------------------------------------------------#')
\# 浅拷贝
out2 = nn.ModuleList([layer for _ in range(6)])
print('浅拷贝不同层的初始化权重：\n',out2[0].weight.data, '\n', out2[1].weight.data)
\# 验证浅拷贝
print('验证浅拷贝：')
input = torch.randn(1,256)
label = torch.ones((1,4), dtype= torch.float32)
\# 取出第一层layer来作为net
net2 = out2[0]
optimizer2 = optim.SGD(net2.parameters(), lr=0.1)
criterion2 = nn.SmoothL1Loss()

\# 迭代
for i in range(1):
    optimizer2.zero_grad()
    res2 = net2(input)
    loss2 = criterion1(res2, label)
    loss2.backward()
    optimizer2.step()
    print('更新后的网络参数：\n',net2.weight.data)
    print('其余层网络参数：\n',out2[1].weight.data)

\#print(id(out1[0]) == id(out1[1]) == id(layer))  # False
\#print(id(out2[0]) == id(out2[1]) == id(layer))  # True`

**Comment by TsingWei** [comment]
酷

**Comment by SlongLiu** [comment]
Thanks for your comments. We do use the shallow copy as we want to share parameters across decoder layers. In our experiments, we found sharing weights of parameters has no negative effects on the final results for our DAB-DETR.

For the ```deepcopy```, we build a new module in every step of a loop other than using the variable of a module. 
We use:
```python
layers = [MLP() for i in range(6)]
```

which is different from:
```python
layer = MLP()
layers = [layer  for i in range(6)]
```

Therefore, I think the implementation in our code is right.

**Comment by wulele2** [comment]
I have verified that your code is correct. Thank you for correcting my bias. Sincerely thank you.
